File: server/question_engine.py
"""Rule-based question-answering engine — Python port of js/engine/questionEngine.js.

This runs entirely server-side now, so the secret character it operates on is never
sent to the browser. `answer_question` is tried first; if it returns matched=False,
app.py falls through to the AI fallback (ai_fallback.py) before finally telling the
player the question wasn't understood.
"""

import re

from answer_types import make_answer
from question_bank import QUESTION_BANK
from schema import NT_BOOKS, OT_BOOKS

_PUNCT_RE = re.compile(r"[?.!,;:'\"()]")
_WS_RE = re.compile(r"\s+")

TIMELINE_CONFIDENCE_GAP = 3

# Words that mark text as an actual question rather than a bare name typed into the
# question box. If any of these lead the input (or it contains "?"), we never treat
# it as an implicit guess — see looks_like_name_guess().
_QUESTION_STARTERS = re.compile(
    r"^(did|do|does|was|were|is|are|am|will|would|should|can|could|have|has|had|"
    r"who|what|when|where|why|how)\b"
)


def normalize(text):
    t = text.lower()
    t = _PUNCT_RE.sub(" ", t)
    t = _WS_RE.sub(" ", t).strip()
    return t


def looks_like_name_guess(raw_text, all_characters):
    """True if `raw_text` isn't phrased as a question at all and exactly matches a
    known character's name or alternate name (anywhere in the full roster, not just
    the secret one — the player is guessing from their own knowledge, not ours).

    Used by app.py to redirect a bare "David" typed into the question box straight
    to the guess flow instead of letting it die as an unparseable question or
    burning an AI call on something that was never a yes/no question."""
    stripped = raw_text.strip()
    if not stripped or "?" in stripped:
        return False
    if _QUESTION_STARTERS.match(stripped.lower()):
        return False

    key = stripped.lower()
    for c in all_characters:
        if c["name"].strip().lower() == key:
            return True
        if any(a.strip().lower() == key for a in c.get("alternate_names", [])):
            return True
    return False


def _try_named_comparison(t, character, all_characters):
    direction = "before" if re.search(r"\bbefore\b", t) else ("after" if re.search(r"\bafter\b", t) else None)
    if not direction:
        return None

    candidates = []
    for c in all_characters:
        if c["id"] == character["id"]:
            continue
        names = [c["name"]] + c.get("alternate_names", [])
        if any(len(n) >= 3 and n.lower() in t for n in names):
            candidates.append(c)
    if len(candidates) != 1:
        return None  # ambiguous or no name found

    other = candidates[0]
    diff = character["timeline_order"] - other["timeline_order"]
    if abs(diff) < TIMELINE_CONFIDENCE_GAP:
        # Names the character the PLAYER typed, never the secret one — safe.
        return make_answer("sometimes", f"Our timelines are too close (or overlap) to say for certain relative to {other['name']}.")

    lived_before = diff < 0
    matches = lived_before if direction == "before" else not lived_before
    return make_answer("yes" if matches else "no")


def answer_question(raw_text, character, all_characters):
    """Returns {"matched": bool, "question_id": str|None, "category": str|None, "answer": dict|None}"""
    t = normalize(raw_text)
    if not t:
        return {"matched": False}

    comparison = _try_named_comparison(t, character, all_characters)
    if comparison:
        return {"matched": True, "question_id": "named_comparison", "category": "Time period", "answer": comparison}

    for q in QUESTION_BANK:
        if q["match"](t):
            return {"matched": True, "question_id": q["id"], "category": q["category"], "answer": q["evaluate"](character)}

    # Cross-testament mention, derived from books_mentioned_in rather than a hand-set flag.
    if re.search(r"mentioned\b.*\bnew testament\b", t) or ("new testament" in t and "mention" in t):
        mentioned = character["testament"] == "NT" or any(b in NT_BOOKS for b in character["books_mentioned_in"])
        return {"matched": True, "question_id": "mentioned_nt", "category": "Identity", "answer": make_answer("yes" if mentioned else "no")}
    if re.search(r"mentioned\b.*\bold testament\b", t) or ("old testament" in t and "mention" in t):
        mentioned = character["testament"] == "OT" or any(b in OT_BOOKS for b in character["books_mentioned_in"])
        return {"matched": True, "question_id": "mentioned_ot", "category": "Identity", "answer": make_answer("yes" if mentioned else "no")}

    # NOTE: deliberately no generic keyword/event fallback — see questionBank.js for why.
    # Anything unmatched here is handed to the AI fallback by app.py, never guessed at locally.
    return {"matched": False}


def check_guess(guess_text, character):
    g = guess_text.strip().lower()
    if not g:
        return False
    if character["name"].lower() == g:
        return True
    return any(a.lower() == g for a in character.get("alternate_names", []))

File: js/engine/questionEngine.js
// Rule-based question-answering engine (no external API).
//
// This is intentionally the only implementation for now, but it is written behind
// a single entry point — `answerQuestion(text, character, allCharacters)` — so a
// future AI-backed answerer can be swapped in later (e.g. try the rule-based
// engine first for speed/reliability, fall back to an LLM call for anything it
// doesn't recognize) without touching game/gameState.js or the UI.

import { QUESTION_BANK } from './questionBank.js';
import { makeAnswer } from './answerTypes.js';
import { OT_BOOKS, NT_BOOKS } from '../data/schema.js';

export function normalize(text) {
  return text
    .toLowerCase()
    .replace(/[?.!,;:'"()]/g, ' ')
    .replace(/\byou\b/g, 'you ')
    .replace(/\s+/g, ' ')
    .trim();
}

// "Did you live before/after <name>?" — resolved via each character's timeline_order.
// A gap smaller than this is treated as "too close to call" rather than guessed at.
const TIMELINE_CONFIDENCE_GAP = 3;

function tryNamedComparison(t, character, allCharacters) {
  const direction = /\bbefore\b/.test(t) ? 'before' : /\bafter\b/.test(t) ? 'after' : null;
  if (!direction) return null;

  const candidates = allCharacters
    .filter((c) => c.id !== character.id)
    .filter((c) => {
      const names = [c.name, ...c.alternate_names].map((n) => n.toLowerCase());
      return names.some((n) => n.length >= 3 && t.includes(n));
    });
  if (candidates.length !== 1) return null; // ambiguous or no name found

  const other = candidates[0];
  const diff = character.timeline_order - other.timeline_order;
  if (Math.abs(diff) < TIMELINE_CONFIDENCE_GAP) {
    // Names the character the PLAYER typed, never the secret one — safe.
    return makeAnswer('sometimes', {
      hint: `Our timelines are too close (or overlap) to say for certain relative to ${other.name}.`,
    });
  }
  const livedBefore = diff < 0;
  const matches = direction === 'before' ? livedBefore : !livedBefore;
  return makeAnswer(matches ? 'yes' : 'no');
}

// NOTE: there is deliberately no generic keyword/event fallback here. Guessing an
// answer from loose text matches against associated_events/people/roles risks
// leaking the character (e.g. a "yes" that's only true for one person in the
// roster is itself a strong clue). Anything the QUESTION_BANK can't confidently
// answer falls through to the AI fallback (server-side), which is held to the
// same never-identify rule — see server/ai_fallback.py.

/**
 * @param {string} rawText - the player's typed question
 * @param {import('../data/schema.js').Character} character - the secret character
 * @param {import('../data/schema.js').Character[]} allCharacters - full roster, used to
 *   resolve "before/after <name>" comparisons
 * @returns {{ matched: boolean, questionId?: string, category?: string, answer?: object }}
 */
export function answerQuestion(rawText, character, allCharacters) {
  const t = normalize(rawText);
  if (!t) return { matched: false };

  const comparison = tryNamedComparison(t, character, allCharacters);
  if (comparison) {
    return { matched: true, questionId: 'named_comparison', category: 'Time period', answer: comparison };
  }

  for (const q of QUESTION_BANK) {
    if (q.match(t)) {
      return { matched: true, questionId: q.id, category: q.category, answer: q.evaluate(character) };
    }
  }

  // Cross-testament mention, derived from books_mentioned_in rather than a hand-set flag.
  if (/mentioned\b.*\bnew testament\b/.test(t) || (/\bnew testament\b/.test(t) && /mention/.test(t))) {
    const mentioned = character.testament === 'NT' || character.books_mentioned_in.some((b) => NT_BOOKS.includes(b));
    return { matched: true, questionId: 'mentioned_nt', category: 'Identity', answer: makeAnswer(mentioned ? 'yes' : 'no') };
  }
  if (/mentioned\b.*\bold testament\b/.test(t) || (/\bold testament\b/.test(t) && /mention/.test(t))) {
    const mentioned = character.testament === 'OT' || character.books_mentioned_in.some((b) => OT_BOOKS.includes(b));
    return { matched: true, questionId: 'mentioned_ot', category: 'Identity', answer: makeAnswer(mentioned ? 'yes' : 'no') };
  }

  return { matched: false };
}

export function checkGuess(guessText, character) {
  const g = guessText.trim().toLowerCase();
  if (!g) return false;
  if (character.name.toLowerCase() === g) return true;
  return character.alternate_names.some((a) => a.toLowerCase() === g);
}
